File: cyberark-upload-files-to-sfe.py
import requests, json, base64, os
from requests.packages.urllib3.exceptions import InsecureRequestWarning
requests.packages.urllib3.disable_warnings(InsecureRequestWarning)

verifySSL = False
headers = {}

def login(user, passwd):
    global session_id, headers
    res = requests.post(base_url + 'WebServices/auth/Cyberark/CyberArkAuthenticationService.svc/Logon', json = { "username": user, "password": passwd }, verify = verifySSL)
    if (res.status_code == 200):
        session_id = (res.json())['CyberArkLogonResult']
        headers = {'Authorization': session_id}
        return True
    else:
        return False
        
def logout():
    res = requests.post(base_url + 'WebServices/auth/Cyberark/CyberArkAuthenticationService.svc/Logoff', headers = headers, verify = verifySSL)
    if (res.status_code == 200):
        return True
    else:
        return False

def upload_file(safe_name, file_path, folder_name):
    file_name = file_path.split('\\')[len(file_path.split('\\'))-1]
    with open(file_path, 'rb') as f:
        res = requests.post(base_url + 'WebServices/API.svc/Upload/Safes/' + safe_name + '/Files/' + file_name + '?folder=' + folder_name, headers = headers, verify = verifySSL, data=f)
    return (res.status_code)

session_id = ''
username = '{username}'
password = '{password}'
host = '{cyberark-sfe-address}'
base_url = 'https://' + host + '/SFE/'
safe_name = '{safe-name}'
folder_path = '/tmp/files/'

if login(username, password):
    print('logged-in successfully.')
    print ('uploading files from \'' + folder_path + '\' to safe \'' + safe_name + '\'...')
    for file in os.listdir(folder_path):
        print ('   uploading \'' + file + '\'... ', end='')
        res = upload_file(safe_name, folder_path + '\/' + file, 'root')
        if res == 201:
            print ('success.')
        elif res == 409:
            print ('failed (file already exists).')
        else:
            print ('failed (' + res + ').')
            
    if logout():
        print('logged-out successfully.')
    
else:
    print('login failed.')

print('\nthank you, come again :)')

File: cyberark-upload-files-to-sfe.php
<?php

    error_reporting(E_ERROR | E_PARSE);
    $session_id = '';
    $username = '{username}';
    $password = '{password}';
    $safe_name = '{safe-name}';
    $host = '{cyberark-sfe-address}';
    $base_url = 'https://'.$host.'/SFE/';
    $folder_path = '{folder_path}';
    $ignore_ssl_issues = false;

    function post($url, $headers, $payload) {
        global $ignore_ssl_issues;
        $options = array('http' => array('method' => 'POST', 'header' => $headers, 'content' => $payload));
        if ($ignore_ssl_issues) $options['ssl'] = array('verify_peer' => false, 'verify_peer_name' => false, 'allow_self_signed' => true);
        $context = stream_context_create($options);
        $response = file_get_contents($url, FALSE, $context);
        if (isset($http_response_header)) return array('status' => explode(' ', $http_response_header[0])[1], 'data' => $response);
    }

    function login($user, $pass) {

        global $base_url, $session_id;
        $url = $base_url.'WebServices/auth/Cyberark/CyberArkAuthenticationService.svc/Logon';
        $payload = array('username' => $user, 'password' => $pass);
        $result = post($url, "Content-Type: application/json\r\nAuthorization: \r\n", json_encode($payload));
        if ($result['status'] == '200') {
            $session_id = json_decode($result['data'], TRUE)['CyberArkLogonResult'];
            return true;
        }
        return false;
    }

    function upload_file($safe_name, $file_name, $folder_name) {
        global $base_url, $folder_path, $session_id;
        $payload = file_get_contents($folder_path.$file_name);
        $url = $base_url.'WebServices/API.svc/Upload/Safes/'.$safe_name.'/Files/'.$file_name.'?folder='.$folder_name;
        $result = post($url, "Authorization: $session_id\r\n", $payload);
        if ($result['status'] == '201') echo 'success.'."\r\n";
        else if ($result['status'] == '409') echo 'failed (file already exists).'."\r\n";
        else echo 'failed ('.$result['status'].').'."\r\n";

    }

    function logout() {
        global $base_url, $session_id;
        $url = $base_url.'WebServices/auth/Cyberark/CyberArkAuthenticationService.svc/Logoff';
        $payload = array();
        return post($url, "Content-Type: application/json\r\nAuthorization: $session_id\r\n", json_encode($payload))['status'] == '200';
    }

    function error_handler($errno, $errstr, $errfile, $errline)
    {
        if (strpos($errstr, 'SSL') !== false) echo 'SSL Error, you may try to set the \'ignore_ssl_issues\' flag to true...'."\r\n";
        return true;
    }

    set_error_handler("error_handler");
    if (!is_dir($folder_path)) echo 'The directory '.$folder_path.' does not exist.'."\r\n";
    else {
        if (!login($username, $password)) echo 'login failed.'."\r\n";
        else {
            echo 'logged-in successfully.'."\r\n";
            echo 'uploading files from \''.$folder_path.'\' to safe \''.$safe_name.'\'...'."\r\n";
            $files = scandir($folder_path);
            foreach($files as $file) {
                if ($file != '.' && $file != '..') {
                    echo '   uploading file \''.$file.'\'... ';
                    upload_file($safe_name, $file, 'root');
                }
            }
            if (logout()) echo 'logged-out successfully.'."\r\n";
        }
    }

?>
